cmd/kube-apiserver/app/server.go

```diff
@@ -103,43 +103,66 @@ func Run(runOptions *options.ServerRunOptions, stopCh <-chan struct{}) error {
 	if err != nil {
 		return err
 	}
-	kubeAPIServer, err := CreateKubeAPIServer(kubeAPIServerConfig, sharedInformers)
+
+	// kubeAPIServer is at the base for now.  This ensures that CustomResourceDefinitions trump TPRs
+	kubeAPIServer, err := CreateKubeAPIServer(kubeAPIServerConfig, genericapiserver.EmptyDelegate, sharedInformers)
 	if err != nil {
 		return err
 	}
 
-	// run the insecure server now, don't block.  It doesn't have any aggregator goodies since authentication wouldn't work
-	if insecureServingOptions != nil {
-		insecureHandlerChain := kubeserver.BuildInsecureHandlerChain(kubeAPIServer.GenericAPIServer.UnprotectedHandler(), kubeAPIServerConfig.GenericConfig)
-		if err := kubeserver.NonBlockingRun(insecureServingOptions, insecureHandlerChain, stopCh); err != nil {
-			return err
-		}
-	}
-
 	// if we're starting up a hacked up version of this API server for a weird test case,
 	// just start the API server as is because clients don't get built correctly when you do this
 	if len(os.Getenv("KUBE_API_VERSIONS")) > 0 {
+		if insecureServingOptions != nil {
+			insecureHandlerChain := kubeserver.BuildInsecureHandlerChain(kubeAPIServer.GenericAPIServer.UnprotectedHandler(), kubeAPIServerConfig.GenericConfig)
+			if err := kubeserver.NonBlockingRun(insecureServingOptions, insecureHandlerChain, stopCh); err != nil {
+				return err
+			}
+		}
+
 		return kubeAPIServer.GenericAPIServer.PrepareRun().Run(stopCh)
 	}
 
 	// otherwise go down the normal path of standing the aggregator up in front of the API server
 	// this wires up openapi
 	kubeAPIServer.GenericAPIServer.PrepareRun()
+
+	// TPRs are enabled and not yet beta, since this these are the successor, they fall under the same enablement rule
+	// Subsequent API servers in between here and kube-apiserver will need to be gated.
+	// These come first so that if someone registers both a TPR and a CRD, the CRD is preferred.
```

Comment:
> What is your rationale for preferring CRD? I'm thinking of a scenario like this:
> 
> TPR already exists and is being used.
> Upgrade to 1.7.
> Accidentally create CRD that conflicts.
> Effectively all your TPR data just got deleted. Outage ensues.
> I would think then it actually makes more sense to let TPR have precedence over CRD. What do you think?

I wanted to prefer the more current API and it makes finalization logic simpler.  Otherwise we'll be considering non-REST API approaches to finalization to avoid deleting all your TPR instances.

We do have a spot that sits above both APIs and could have knowledge of them to avoid collisions.  An admission plugin could be written to prevent conflicts between these.